docs/architecture-decisions/adr004-module-export-structure.md

```diff
@@ -0,0 +1,71 @@
+# ADR004: Module Export Structure
+
+| Created    | Status |
+| ---------- | ------ |
+| 2020-05-27 | Open   |
+
+## Context
+
+With a growing number of exports of packages like `@backstage/core`, it is becoming more and more difficult to answer questions such as
+
+> Is the export in this module also exported by the package?
+
+or
+
+> What is exported from this directory?
+
+We currently do not use any pattern for how to structure exports. There is a mix of package-level re-exports deep into the directory tree, shallow re-exports for each directory, exports using `*` and explicit lists of each symbol, etc.
+The mix and lack of predictability makes it difficult to reason about the boundaries of a module, and for example knowing whether is is safe to export a symbol in a given file.
+
+## Decision
+
+We will make each exported symbol traceable through index files all the way down to the root of the package, `src/index.ts`. Each index file will only re-export from its own immediate directory children, and only index files will have re-exports. This gives a file tree similar to this:
+
+```text
+index.ts
+components/index.ts
+          /ComponentX/index.ts
+                     /ComponentX.tsx
+                     /SubComponentY.tsx
+lib/index.ts
+   /UtilityX/index.ts
+            /UtilityX.ts
+            /helper.ts
+```
+
+To check whether for example `SubComponentY` is exported from the package, it should be possible to traverse the index files towards the root, starting at the adjacent one. If there is any index file that doesn't export the previous one, the symbol is not publicly exported. For example, if `components/ComponentX/index.ts` exports `SubComponentY`, but `components/index.ts` does not re-export `./ComponentX`, one should be certain that `SubComponentY` is not exported outside the package. This rule would be broken if for example the root `index.ts` re-exports `./components/ComponentX`
```

Comment:
`components/index.ts` does not re-export `./ComponentX` -> `components/index.ts` does not re-export `./SubComponentY` ?

or am I understanding this wrong?